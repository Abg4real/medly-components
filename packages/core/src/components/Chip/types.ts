import { HTMLProps, Omit, WithThemeProp } from '@medly-components/utils';

type ButtonProps = Omit<HTMLProps<HTMLButtonElement>, 'type'>;

export interface StyledProps extends ButtonProps, WithThemeProp {
    variant?: 'solid' | 'flat' | 'outlined';
    color?: string;
    as?: any;
    to?: string;
}

export interface Props extends StyledProps {
    label: string;
    onClick?: () => void;
    onDelete?: () => void;
}
