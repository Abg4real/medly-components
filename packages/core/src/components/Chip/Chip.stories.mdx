import { Chip } from './Chip';
import { defaultTheme } from '@medly-components/theme';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Chip.stories';

<Meta title="Core/Chip" component={Chip} />

# Chip

### Basic Chip

<Preview withToolbar>
    <Story name="Basic">{stories.Basic}</Story>
</Preview>

### Props

<Props of={Chip} />

### ChipTheme

<Props of={stories.ThemeInterface} />

### ChipTheme Defaults

<code>{JSON.stringify(defaultTheme.chip, null, 2)}</code>
