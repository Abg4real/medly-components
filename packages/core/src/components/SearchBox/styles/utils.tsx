import { SvgIcon } from '@medly-components/icons';
import { css } from '@medly-components/utils';
import { Props } from '../types';

export const getSearchBoxSize = ({ theme, size }: Props) => theme.searchBox.sizes[size].height;

export const getBorderAndBoxShadow = (color: string, boxShadow: string) => {
    return css`
        border-color: ${color};
        box-shadow: ${boxShadow};
    `;
};

export const getIconWrapperStyle = ({ theme, size }: Props) => {
    return css`
        width: ${getSearchBoxSize};
        min-width: ${getSearchBoxSize};
        height: auto;
        display: inline-flex;
        align-items: center;
        justify-content: center;
        ${SvgIcon} {
            cursor: pointer;
            font-size: ${theme.searchBox.sizes[size].iconSize};
        }
    `;
};
