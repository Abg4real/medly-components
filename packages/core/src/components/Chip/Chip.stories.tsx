import { ChipTheme, defaultTheme } from '@medly-components/theme';
import { action } from '@storybook/addon-actions';
import { boolean, color, select, text } from '@storybook/addon-knobs';
import React from 'react';
import { Chip } from './Chip';
import { Props } from './types';

const variant: Array<Props['variant']> = ['solid', 'flat', 'outlined'];

export const ThemeInterface = (props: ChipTheme): any => null;

export const Basic = () => (
    <Chip
        label={text('Label', 'Demo')}
        variant={select('Variant', variant, 'solid')}
        onClick={action('Chip Clicked')}
        onDelete={action('Delete Clicked')}
        disabled={boolean('Disabled', false)}
        color={color('Color', defaultTheme.chip.defaultColor)}
    />
);
