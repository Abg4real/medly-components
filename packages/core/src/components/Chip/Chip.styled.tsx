import { SvgIcon } from '@medly-components/icons';
import { centerAligned, css, styled } from '@medly-components/utils';
import Text from '../Text';
import { StyledProps } from './types';

const getColor = ({ theme, color, disabled }: StyledProps) => (disabled ? theme.button.disabled : color || theme.chip.defaultColor);

const outlined = () => css`
    border: 1px solid ${getColor};
    color: ${getColor};
    ${SvgIcon} {
        * {
            fill: ${getColor};
        }
    }
`;

const solid = () => css`
    border: 1px solid ${getColor};
    background-color: ${getColor};
    color: ${({ theme }) => theme.colors.white};
    ${SvgIcon} {
        * {
            fill: ${({ theme }) => theme.colors.white};
        }
    }
`;

const flat = () => css`
    color: ${getColor};
    ${SvgIcon} {
        * {
            fill: ${getColor};
        }
    }
`;

export const Chip = styled('button')<StyledProps>`
    background-color: transparent;
    border: none;
    border-radius: 20px;
    padding: 5px 10px;
    user-select: none;
    text-decoration: none;

    &:hover {
        cursor: ${({ onClick }) => (onClick ? 'pointer' : 'default')};
    }

    &:focus {
        outline: none;
    }

    ${props => props.variant === 'flat' && flat()};
    ${props => props.variant === 'outlined' && outlined()};
    ${props => props.variant === 'solid' && solid()};

    &:disabled {
        cursor: not-allowed;
    }

    ${Text.Style} + ${SvgIcon}{
        cursor: pointer;
        margin-left: 10px;
    }

    ${centerAligned()}
`;
