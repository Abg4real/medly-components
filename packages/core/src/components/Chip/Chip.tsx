import { ClearIcon } from '@medly-components/icons';
import { WithStyle } from '@medly-components/utils';
import React, { SFC, useCallback } from 'react';
import Text from '../Text';
import * as Styled from './Chip.styled';
import { Props } from './types';

export const Chip: SFC<Props> & WithStyle = React.memo(
    React.forwardRef((props, ref) => {
        const { label, onDelete, ...restProps } = props;

        const handleDelete = useCallback(
            (e: React.MouseEvent<HTMLOrSVGElement>) => {
                e.stopPropagation();
                !restProps.disabled && onDelete && onDelete();
            },
            [restProps.disabled, onDelete]
        );

        return (
            <Styled.Chip ref={ref} {...restProps}>
                <Text>{label}</Text>
                {onDelete && <ClearIcon size="XS" onClick={handleDelete} />}
            </Styled.Chip>
        );
    })
);

Chip.defaultProps = {
    variant: 'solid'
};
Chip.displayName = 'Chip';
Chip.Style = Styled.Chip;
