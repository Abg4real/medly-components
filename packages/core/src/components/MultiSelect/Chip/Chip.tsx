import { ClearIcon } from '@medly-components/icons';
import { WithStyle } from '@medly-components/utils';
import React, { SFC, useCallback } from 'react';
import Text from '../../Text';
import * as Styled from './Chip.styled';
import { Props } from './types';

export const Chip: SFC<Props> & WithStyle = React.memo(
    React.forwardRef((props, ref) => {
        const { id, label, onClear, ...restProps } = props,
            chipId = id ? id.replace(' ', '') : 'medly-options-chip';

        const onClearHandler = useCallback(
            (e: React.MouseEvent<HTMLOrSVGElement>) => {
                e.stopPropagation();
                !restProps.disabled && onClear && onClear(label);
            },
            [restProps.disabled, onClear]
        );

        return (
            <Styled.Chip ref={ref} id={chipId} {...restProps}>
                <Text textWeight="Medium" textVariant={restProps.size === 'S' ? 'body2' : 'body1'}>
                    {label}
                </Text>
                <ClearIcon id={`${chipId}-clear`} size="S" onClick={onClearHandler} />
            </Styled.Chip>
        );
    })
);

Chip.displayName = 'Chip';
Chip.Style = Styled.Chip;
Chip.defaultProps = {
    state: 'default',
    variant: 'outlined'
};
