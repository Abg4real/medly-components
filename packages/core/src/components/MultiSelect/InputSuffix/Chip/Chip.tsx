import { ClearIcon } from '@medly-components/icons';
import { WithStyle } from '@medly-components/utils';
import React, { FC, useCallback } from 'react';
import Text from '../../../Text';
import * as Styled from './Chip.styled';
import { Props } from './types';

export const Chip: FC<Props> & WithStyle = React.memo(({ id, label, onClear, ...restProps }) => {
    const onClickHandler = useCallback((e: React.MouseEvent) => e.stopPropagation(), []),
        onClearHandler = useCallback(
            (e: React.MouseEvent<HTMLOrSVGElement>) => {
                e.stopPropagation();
                !restProps.disabled && onClear && onClear(label);
            },
            [restProps.disabled, onClear]
        );

    return (
        <Styled.Chip id={id} {...restProps} onClick={onClickHandler}>
            <Text textWeight="Medium" textVariant={restProps.size === 'S' ? 'body2' : 'body1'}>
                {label}
            </Text>
            {!restProps.hideClearIcon && <ClearIcon id={`${id}-clear`} onClick={onClearHandler} />}
        </Styled.Chip>
    );
});

Chip.displayName = 'Chip';
Chip.Style = Styled.Chip;
Chip.defaultProps = {
    state: 'default',
    variant: 'outlined'
};
