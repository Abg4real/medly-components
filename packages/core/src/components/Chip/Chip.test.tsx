import { TestUtils } from '@medly-components/utils';
import React from 'react';
import Chip from './Chip';

describe('Chip component', () => {
    it('should render properly with flat variant', () => {
        const { container } = TestUtils.render(<Chip variant="flat" color="yellow" label="Flat Chip" />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly with outlined variant', () => {
        const { container } = TestUtils.render(<Chip variant="outlined" color="secondary" label="Outlined Chip" />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly with solid variant', () => {
        const { container } = TestUtils.render(<Chip variant="solid" label="Solid Chip" />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly when it is disabled', () => {
        const { container } = TestUtils.render(<Chip disabled variant="solid" label="Solid Chip" />);
        expect(container).toMatchSnapshot();
    });

    it('should call onClick when click on it', () => {
        const mockOnClick = jest.fn();

        const { getByText } = TestUtils.render(<Chip variant="flat" color="yellow" onClick={mockOnClick} label="Flat Chip" />);
        TestUtils.fireEvent.click(getByText('Flat Chip'));
        expect(mockOnClick).not.toBeCalled();
    });

    it('should call onDelete when click on clear icon', () => {
        const mockOnClick = jest.fn(),
            mockOnDelete = jest.fn();

        const { container } = TestUtils.render(
            <Chip variant="flat" onClick={mockOnClick} onDelete={mockOnDelete} color="yellow" label="Flat Chip" />
        );
        TestUtils.fireEvent.click(container.querySelector('svg'));
        expect(mockOnDelete).toBeCalled();
        expect(mockOnClick).not.toBeCalled();
    });
});
