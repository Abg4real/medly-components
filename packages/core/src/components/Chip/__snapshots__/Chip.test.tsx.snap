// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip component should render properly when it is disabled 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c3 .c1 + .c6 {
  cursor: pointer;
  margin-left: 10px;
}

.c4 .c1 + .c6 {
  cursor: pointer;
  margin-left: 10px;
}

.c5 .c1 + .c6 {
  cursor: pointer;
  margin-left: 10px;
}

.c0 {
  background-color: transparent;
  border: none;
  border-radius: 20px;
  padding: 5px 10px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: 1px solid #AAAAAA;
  background-color: #AAAAAA;
  color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0 .c6 * {
  fill: #ffffff;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 .c1 + .c6 {
  cursor: pointer;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    <span
      class="c1 c2"
    >
      Solid Chip
    </span>
  </button>
</div>
`;

exports[`Chip component should render properly with flat variant 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c0 {
  background-color: transparent;
  border: none;
  border-radius: 20px;
  padding: 5px 10px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: yellow;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0 .c3 * {
  fill: yellow;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 .c1 + .c3 {
  cursor: pointer;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
    color="yellow"
  >
    <span
      class="c1 c2"
    >
      Flat Chip
    </span>
  </button>
</div>
`;

exports[`Chip component should render properly with outlined variant 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c3 .c1 + .c4 {
  cursor: pointer;
  margin-left: 10px;
}

.c0 {
  background-color: transparent;
  border: none;
  border-radius: 20px;
  padding: 5px 10px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: 1px solid secondary;
  color: secondary;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0 .c4 * {
  fill: secondary;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 .c1 + .c4 {
  cursor: pointer;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
    color="secondary"
  >
    <span
      class="c1 c2"
    >
      Outlined Chip
    </span>
  </button>
</div>
`;

exports[`Chip component should render properly with solid variant 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c3 .c1 + .c5 {
  cursor: pointer;
  margin-left: 10px;
}

.c4 .c1 + .c5 {
  cursor: pointer;
  margin-left: 10px;
}

.c0 {
  background-color: transparent;
  border: none;
  border-radius: 20px;
  padding: 5px 10px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: 1px solid #012040;
  background-color: #012040;
  color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0 .c5 * {
  fill: #ffffff;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 .c1 + .c5 {
  cursor: pointer;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1 c2"
    >
      Solid Chip
    </span>
  </button>
</div>
`;
